### PR TITLE
(CDAP-4173) Enhance MapReduceContext.setInput

### PR DESCRIPTION
**Pull request by chtyim**
- Allow dataset arguments to be passed to setInput
- Add a new setInput that accepts InputFormatProvider
- Deprecate setInput(String name, Dataset dataset) method
  - It can be replaced with the one that accept arguments or the one accepts InputFormatProvider
- Code refactoring/simplification to use InputFormatProvider to configure MR job

### CODE REVIEW

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/AppWithTxAware.java [hunk 1]
```diff
@@ -62,7 +62,7 @@ public void beforeSubmit(MapReduceContext context) throws Exception {
       job.setNumReduceTasks(0);
       FileOutputFormat.setOutputPath(job, new Path(context.getRuntimeArguments().get("outputPath")));
       PedanticTxAware input = context.getDataset("pedanticTxAware", ImmutableMap.of("value", "1"));
-      context.setInput("pedanticTxAware", input);
+      context.setInput("pedanticTxAware", ImmutableMap.of("value", "1"));
```
Comment (gokulavasan) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/AppWithTxAware.java]: Do we need to specify the properties twice? Once while getting the dataset and again while setting the input of the MapReduce job?

**Comment by gokulavasan** [comment]
One question. Rest LGTM :+1:

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/AppWithTxAware.java (hunk 1 above)
Comment (chtyim) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/AppWithTxAware.java]: Yes. A dataset is unique by name and properties

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceTaskContext.java [hunk 2]
```diff
@@ -98,7 +98,7 @@ public BasicMapReduceTaskContext(Program program,
                                    DatasetFramework dsFramework,
                                    @Nullable PluginInstantiator pluginInstantiator,
                                    Map<String, File> localizedResources) {
-    super(program, runId, runtimeArguments, datasets,
+    super(program, runId, runtimeArguments, ImmutableSet.<String>of(),
```
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceTaskContext.java]: Why this change?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/AppWithMapReduceUsingRuntimeDatasets.java [hunk 3]
```diff
@@ -84,26 +83,21 @@ public void beforeSubmit(MapReduceContext context) throws Exception {
       String outputName = runtimeArgs.get(OUTPUT_NAME);
       String outputPath = runtimeArgs.get(OUTPUT_PATH);
 
-      FileSet input;
-      FileSet output;
       if (inputName.equals(outputName)) {
```
Comment (anwar6953) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/AppWithMapReduceUsingRuntimeDatasets.java]: It seems simpler to just replace the if and else blocks using the code from the else case.
Its the same code, only uses 1 less hashmap, but makes the code messier, harder to read.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceTaskContext.java (hunk 2 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceTaskContext.java]: The old code always pass in empty set.

Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceTaskContext.java]: :+1:

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/AppWithMapReduceUsingRuntimeDatasets.java (hunk 3 above)
Comment (chtyim) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/AppWithMapReduceUsingRuntimeDatasets.java]: I was not intended to change the current code logic, but to merely use the new API instead of the now deprecated one.

Comment (chtyim) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/AppWithMapReduceUsingRuntimeDatasets.java]: Anyway, simplified the code.

**Comment by anwar6953** [comment]
LGTM

**Comment by chtyim** [comment]
Build passed: http://builds.cask.co/browse/CDAP-DUT3075-9

**Comment by chtyim** [comment]
Squashed commits.
